lib/data_stores/mysql_advanced_data_store.py

```diff
@@ -0,0 +1,636 @@
+  """A mysql based data store."""
+
+  POOL = None
+  SYSTEM_TABLE = "system"
+
+  def __init__(self):
+    # Use the global connection pool.
+    if MySQLAdvancedDataStore.POOL is None:
+      MySQLAdvancedDataStore.POOL = ConnectionPool()
+    self.pool = self.POOL
+
+    self.to_insert = []
+    self._CalculateAttributeStorageTypes()
+    self.database_name = config_lib.CONFIG["Mysql.database_name"]
+    self.lock = threading.Lock()
+
+    super(MySQLAdvancedDataStore, self).__init__()
+
+  def Initialize(self):
+    try:
+      self._ExecuteQuery("desc `aff4`")
+    except MySQLdb.Error:
+      self.RecreateTables()
+
+  def DropTables(self):
+    #Drop all existing tables
+    rows = self._ExecuteQuery("select table_name from information_schema.tables "
+                              "where table_schema='%s'" % self.database_name)
+    for row in rows:
+      self._ExecuteQuery("drop table `%s`" % row["table_name"])
+
+  def RecreateTables(self):
+    """Drops the tables and creates a new ones."""
+    self.DropTables()
+
+    self._CreateTables()
+
+  def Transaction(self, subject, lease_time=None, token=None):
+    return MySQLTransaction(self, subject, lease_time=lease_time, token=token)
+
+  def Size(self):
+    query = ("SELECT table_schema, Sum(data_length + index_length) `size` "
+             "FROM information_schema.tables "
+             "WHERE table_schema = \"%s\" GROUP by table_schema" %
+             self.database_name)
+
+    result = self._ExecuteQuery(query, [])
+    if len(result) != 1:
+      return -1
+    return int(result[0]["size"])
+
+  def DeleteAttributes(self, subject, attributes, start=None, end=None, sync=None, token=None):
+    """Remove some attributes from a subject."""
+    _ = sync  # Unused
+    self.security_manager.CheckDataStoreAccess(token, [subject], "w")
+    if not attributes:
+      return
+
+    for attribute in attributes:
+      timestamp = self._MakeTimestamp(start, end)
+      predicate = utils.SmartUnicode(attribute)
+      query, args = self._BuildQuery("DELETE", subject, predicate, timestamp)
+      self._ExecuteQuery(query, args)
+
+  def DeleteAttributesRegex(self, subject, regexes, token=None):
+    self.security_manager.CheckDataStoreAccess(token, [subject], "w")
+
+    for regex in regexes:
+      query, args = self._BuildQuery("DELETE", subject, regex, is_regex=True)
+      self._ExecuteQuery(query, args)
+
+  def DeleteSubject(self, subject, token=None, sync=False):
+    _ = sync
+    self.security_manager.CheckDataStoreAccess(token, [subject], "w")
+
+    query, args = self._BuildQuery("DELETE", subject)
+    self._ExecuteQuery(query, args)
+
+  def ResolveMulti(self, subject, predicates, token=None, timestamp=None):
+    """Resolves multiple predicates at once for one subject."""
+    self.security_manager.CheckDataStoreAccess(
+        token, [subject], self.GetRequiredResolveAccess(predicates))
+
+    for predicate in predicates:
```

Comment:
The original code did, but I suspect that the MySQL engine does it considerably faster due to the indexes being more efficient than the python loops.  My guess would be that the overhead of using the connection pool to send multiple queries with small exact results is going to be lower than sending a single query with substantially larger results and then having to do the sorting and grouping in python.  Without the connection pool the round trip would probably be a larger factor.

The benchmark tests can't really test this properly due to the data and access patterns not quite lining up with an operational instance and trying to accurately test it live was painfully difficult.  I can say that during earlier testing the bulk queries caused by this were routinely taking much longer and showing up in the slow queries log and then also getting sorted in python, but that doesn't necessarily mean that the aggregate of individual faster queries is better or that other changes are responsible for the large scale performance improvement.